File: src/ruby/sort_insertion_sort.rb
# frozen_string_literal: true

########################
# business logic start #
########################

# insertion sort!
#
# docs: https://en.wikipedia.org/wiki/Insertion_sort
#
# Insertion sort steps through an input list, using it to grow an output list.
# On every step, it sorts the current element into its proper location on the
# output list. It continues until the input list is empty.

def do_sorting(input_list)
  insertion_sort(input_list)
end

def insertion_sort(input_list)
  output_list = []

  input_list.each_with_index do |element, idx|
    output_list.append(element)
    output_list = sort_element_at_index(output_list, element, idx)
  end

  output_list
end

def sort_element_at_index(input_list, element, idx)
  # puts '--- sort_element_at_index ---'
  # puts "element => #{element}"
  # puts "index => #{idx}"
  output_list = input_list
  target_index = idx

  while (target_index != 0) && (element < output_list[target_index - 1])
    output_list.swap(target_index)
    target_index -= 1
    # puts "\ttarget_index => #{target_index}"
    # puts "\toutput_list => #{output_list}"
  end

  output_list
end

# add swap method to array
class Array
  def swap(idx)
    self[idx], self[idx - 1] = self[idx - 1], self[idx]
  end
end

######################
# business logic end #
######################

# 👇🏽 copy pasted helpers

input_file_path = ENV['INPUT_PATH']
input_file = File.readlines(input_file_path)

sorted_data = do_sorting(input_file)

output_file_path = ENV['OUTPUT_PATH']
File.open(output_file_path, 'wb') { |f| f.write(sorted_data.join('')) }

File: src/js/sort_insertion_sort.js
var fs = require("fs");

// business logic start //

// insertion sort - https://en.wikipedia.org/wiki/Insertion_sort
// Steps through the input, growing a sorted output by inserting each element in place.

function doSorting(inputList) {
  return insertionSort(inputList);
}

function insertionSort(inputList) {
  outputList = [];

  inputList.forEach((element, idx) => {
    outputList.push(element);
    outputList = sortElementAtIndex(outputList, element, idx);
  });

  return outputList;
}

function sortElementAtIndex(inputList, element, idx) {
  target_index = idx;
  outputList = inputList;

  while (target_index != 0 && element < outputList[target_index - 1]) {
    swapWithPrevious(outputList, target_index);
    target_index -= 1;
  }

  return outputList;
}

function swapWithPrevious(list, idx) {
  tmp = list[idx - 1];
  list[idx - 1] = list[idx];
  list[idx] = tmp;
  return list;
}

// business logic end //

// 👇🏽 copy pasted helpers

// read input file
let inputFilePath = process.env.INPUT_PATH;
let inputString = fs.readFileSync(inputFilePath, "utf8");
let inputStringSplit = inputString.split(/\n/);

// clean input data
if (inputStringSplit[inputStringSplit.length - 1] === "") {
  inputStringSplit = inputStringSplit.slice(0, inputStringSplit.length - 1);
}

// do sorting
sortedData = doSorting(inputStringSplit);

// write output file
let outputFilePath = process.env.OUTPUT_PATH;
let outputString = sortedData.join("\n");
outputString += "\n"; // trailing newline
fs.writeFileSync(outputFilePath, outputString);
